File: app/src/main/java/com/taylab/fleetrunner/net/CollectorClient.kt
package com.taylab.fleetrunner.net

import com.taylab.fleetrunner.protocol.FleetJson
import com.taylab.fleetrunner.protocol.JobSpec
import com.taylab.fleetrunner.protocol.RegisterPost
import com.taylab.fleetrunner.protocol.ResultPost
import com.taylab.fleetrunner.protocol.intParam
import com.taylab.fleetrunner.protocol.stringParam
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.encodeToString
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.security.MessageDigest
import java.util.concurrent.TimeUnit

/** Blocking HTTP client for the collector; call from Dispatchers.IO. */
class CollectorClient(baseUrl: String) {

    private val base = baseUrl.trimEnd('/')
    private val json = "application/json".toMediaType()
    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        // The collector holds next-job long-polls for ~25 s; stay well above that.
        .readTimeout(40, TimeUnit.SECONDS)
        .build()

    fun register(post: RegisterPost) {
        val req = Request.Builder()
            .url("$base/devices/register")
            .post(FleetJson.encodeToString(post).toRequestBody(json))
            .build()
        http.newCall(req).execute().use { res ->
            if (!res.isSuccessful) throw IOException("register failed: HTTP ${res.code}")
        }
    }

    /** Long-polls for work; null when the poll expired with no job (HTTP 204). */
    fun nextJob(deviceId: String): JobSpec? {
        val req = Request.Builder().url("$base/devices/$deviceId/next-job").build()
        http.newCall(req).execute().use { res ->
            return when {
                res.code == 204 -> null
                res.isSuccessful -> FleetJson.decodeFromString<JobSpec>(res.body!!.string())
                else -> throw IOException("next-job failed: HTTP ${res.code}")
            }
        }
    }

    /** Streams an artifact to [dest], verifying its content hash before the rename. */
    fun downloadArtifact(sha256: String, dest: File) {
        val req = Request.Builder().url("$base/artifacts/$sha256").build()
        http.newCall(req).execute().use { res ->
            if (!res.isSuccessful) throw IOException("artifact $sha256 failed: HTTP ${res.code}")
            val tmp = File(dest.parentFile, "${dest.name}.part")
            val digest = MessageDigest.getInstance("SHA-256")
            tmp.outputStream().use { out ->
                val src = res.body!!.byteStream()
                val buf = ByteArray(1 shl 16)
                while (true) {
                    val n = src.read(buf)
                    if (n < 0) break
                    digest.update(buf, 0, n)
                    out.write(buf, 0, n)
                }
            }
            val got = digest.digest().joinToString("") { "%02x".format(it) }
            if (got != sha256) {
                tmp.delete()
                throw IOException("artifact hash mismatch: wanted $sha256 got $got")
            }
            if (!tmp.renameTo(dest)) throw IOException("rename failed for $dest")
        }
    }

    /** Uploads raw bytes to the artifact store; returns the sha256. */
    fun uploadArtifact(bytes: ByteArray, name: String): String {
        val req = Request.Builder()
            .url("$base/artifacts")
            .header("x-artifact-name", name)
            .post(bytes.toRequestBody("application/octet-stream".toMediaType()))
            .build()
        http.newCall(req).execute().use { res ->
            if (!res.isSuccessful) throw IOException("artifact upload failed: HTTP ${res.code}")
            val body = FleetJson.parseToJsonElement(res.body!!.string()) as JsonObject
            return body.stringParam("sha256") ?: throw IOException("upload response missing sha256")
        }
    }

    fun publishEvent(topic: String, payloadJson: String) {
        val req = Request.Builder()
            .url("$base/events/$topic")
            .post(payloadJson.toRequestBody(json))
            .build()
        http.newCall(req).execute().use { res ->
            if (!res.isSuccessful) throw IOException("publish failed: HTTP ${res.code}")
        }
    }

    data class Event(val id: Long, val payload: JsonObject)

    /** Long-polls the next event after [after]; null when the poll expired. */
    fun pollEvent(topic: String, after: Long): Event? {
        val req = Request.Builder().url("$base/events/$topic/poll?after=$after").build()
        http.newCall(req).execute().use { res ->
            return when {
                res.code == 204 -> null
                res.isSuccessful -> {
                    val body = FleetJson.parseToJsonElement(res.body!!.string()) as JsonObject
                    Event(
                        id = body.intParam("id", 0).toLong(),
                        payload = body["payload"] as? JsonObject ?: JsonObject(emptyMap()),
                    )
                }
                else -> throw IOException("poll failed: HTTP ${res.code}")
            }
        }
    }

    fun postResult(row: ResultPost) {
        val req = Request.Builder()
            .url("$base/results")
            .post(FleetJson.encodeToString(row).toRequestBody(json))
            .build()
        http.newCall(req).execute().use { res ->
            if (!res.isSuccessful) throw IOException("results failed: HTTP ${res.code}")
        }
    }
}

File: FleetRunner/CollectorClient.swift
import Foundation

enum CollectorError: Error, LocalizedError {
    case http(Int, String)
    var errorDescription: String? {
        if case let .http(code, path) = self { return "\(path) failed: HTTP \(code)" }
        return nil
    }
}

/// Async HTTP client for the collector.
final class CollectorClient: Sendable {
    private let base: URL
    private let session: URLSession

    init(baseURL: URL) {
        self.base = baseURL
        let config = URLSessionConfiguration.ephemeral
        // The collector holds next-job long-polls for ~25 s; stay well above.
        config.timeoutIntervalForRequest = 40
        self.session = URLSession(configuration: config)
    }

    @discardableResult
    private func post<T: Encodable>(_ path: String, body: T) async throws -> Data {
        var req = URLRequest(url: base.appendingPathComponent(path))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = try FleetJSON.encoder.encode(body)
        let (data, res) = try await session.data(for: req)
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(code) else { throw CollectorError.http(code, path) }
        return data
    }

    func register(_ body: RegisterPost) async throws {
        try await post("devices/register", body: body)
    }

    /// Long-polls for work; nil when the poll expired with no job (HTTP 204).
    func nextJob(deviceId: String) async throws -> JobSpec? {
        let url = base.appendingPathComponent("devices/\(deviceId)/next-job")
        let (data, res) = try await session.data(from: url)
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        if code == 204 { return nil }
        guard code == 200 else { throw CollectorError.http(code, "next-job") }
        return try FleetJSON.decoder.decode(JobSpec.self, from: data)
    }

    func postResult(_ row: ResultPost) async throws {
        try await post("results", body: row)
    }

    /// Posts a beacon and returns the collector's `lease_renewed` flag: false
    /// means the claim on the beacon's job is gone — cancelled from the
    /// dashboard, or swept for a missed lease — and the job should stop.
    ///
    /// Only an explicit false in a 2xx body says that. An absent field reads as
    /// true, so a collector that doesn't send one changes nothing here, and a
    /// non-2xx response or a transport failure throws rather than quietly
    /// looking like a cancel.
    func postBeacon(_ row: ResultPost) async throws -> Bool {
        let data = try await post("results", body: row)
        struct Ack: Decodable { let leaseRenewed: Bool? }
        return (try? FleetJSON.decoder.decode(Ack.self, from: data))?.leaseRenewed ?? true
    }

    /// Uploads raw bytes to the artifact store; returns the sha256.
    func uploadArtifact(_ data: Data, name: String) async throws -> String {
        var req = URLRequest(url: base.appendingPathComponent("artifacts"))
        req.httpMethod = "POST"
        req.setValue("application/octet-stream", forHTTPHeaderField: "content-type")
        req.setValue(name, forHTTPHeaderField: "x-artifact-name")
        let (body, res) = try await session.upload(for: req, from: data)
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 201 else { throw CollectorError.http(code, "artifacts") }
        struct R: Decodable { let sha256: String }
        return try JSONDecoder().decode(R.self, from: body).sha256
    }

    func publishEvent(topic: String, payload: [String: Any]) async throws {
        var req = URLRequest(url: base.appendingPathComponent("events/\(topic)"))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = try JSONSerialization.data(withJSONObject: payload)
        let (_, res) = try await session.data(for: req)
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 201 else { throw CollectorError.http(code, "events") }
    }

    struct Event { let id: Int; let payload: [String: Any] }

    /// Long-polls the next event after `after`; nil when the poll expired.
    func pollEvent(topic: String, after: Int) async throws -> Event? {
        let url = base.appendingPathComponent("events/\(topic)/poll")
            .appending(queryItems: [URLQueryItem(name: "after", value: String(after))])
        let (data, res) = try await session.data(from: url)
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        if code == 204 { return nil }
        guard code == 200,
              let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let id = obj["id"] as? Int else { throw CollectorError.http(code, "events/poll") }
        return Event(id: id, payload: obj["payload"] as? [String: Any] ?? [:])
    }
}
